### PR TITLE
Enha: Exceptions not handled by user code should be `unhandled: true`

### LINKED ISSUE #456: Exceptions not handled by user code should be `unhandled: true`
Currently all exceptions are marked as `handled`.
This decision was made because exceptions in Flutter don't crash the app and thus sessions should not be marked as crashed.

Though we want to mark exceptions as unhandled to indicate that an exception caught by the root zone handler was not handled by the user.

There should be a way to differentiate.

This issue is a result of a [discussion on Discord](https://discord.com/channels/621778831602221064/654883298736340994/840233050905903186).

### PR DESCRIPTION
**Pull request by denrase**
## :scroll: Description

Closes #456

Always set to unhandled, as there are no handlers provided:

- IsolateErrorIntegration
- RunZonedGuardedIntegration

Set to unhandled if there are no user provided handlers:

- FlutterErrorIntegration
- OnErrorIntegration

Verified in tests that handled is null on the mechanism. Technically those should also be unhandled, as users are not manually calling the Sentry SDK, but failed requests are something so common and should probably not be unhandled?

- IoExceptionEventProcessor
- FailedRequestClient
- FailedRequestInterceptor

## :green_heart: How did you test it?

Added unit tests.

## :pencil: Checklist

- [x] I reviewed submitted code
- [x] I added tests to verify changes
- [ ] I updated the docs if needed
- [x] All tests passing
- [x] No breaking changes

## :crystal_ball: Next steps

Review every individually created mechanism if the `handled` flag makes sense.

### CODE REVIEW

File: dart/lib/src/isolate_error_integration.dart [hunk 1]
```diff
@@ -66,8 +66,9 @@ Future<void> handleIsolateError(
       stackTrace: stackTrace == null ? null : StackTrace.fromString(stackTrace),
     );
 
-    //  Isolate errors don't crash the App.
-    final mechanism = Mechanism(type: 'isolateError', handled: true);
+    //  Isolate errors don't crash the app, but are set to handled false, as
+    // as users did not handle them manually
```
Comment (marandaneto) [inline dart/lib/src/isolate_error_integration.dart]:
```suggestion
    //  Isolate errors don't crash the app, but are set to handled false, as
    // users did not handle them manually
```

File: flutter/lib/src/integrations/flutter_error_integration.dart [hunk 2]
```diff
@@ -59,7 +59,8 @@ class FlutterErrorIntegration extends Integration<SentryFlutterOptions> {
         // FlutterError doesn't crash the App.
         final mechanism = Mechanism(
           type: 'FlutterError',
-          handled: true,
+          handled: _defaultOnError != null &&
+              _defaultOnError != FlutterError.presentError,
```
Comment (marandaneto) [inline flutter/lib/src/integrations/flutter_error_integration.dart]:
```suggestion
          handled: false,
```

Not sure why you did this conditional if, mind explaining?

Comment (denrase) [reply flutter/lib/src/integrations/flutter_error_integration.dart]: Thought that users could provide their own callback to `FlutterError.onError`, which we respect and also call, so I assumed that in this case, we'd mark it as handled.

File: flutter/lib/src/integrations/on_error_integration.dart [hunk 3]
```diff
@@ -42,7 +42,7 @@ class OnErrorIntegration implements Integration<SentryFlutterOptions> {
     _defaultOnError = wrapper.onError;
 
     _integrationOnError = (Object exception, StackTrace stackTrace) {
-      final handled = _defaultOnError?.call(exception, stackTrace) ?? true;
+      final handled = _defaultOnError?.call(exception, stackTrace) ?? false;
```
Comment (marandaneto) [inline flutter/lib/src/integrations/on_error_integration.dart]: I'm not sure if we should use the result of this function as the sentry `handled` flag.
I believe we should just use `false` since the concept is different.
About fallback to `true` or `false`, I'd check whats the default value for `https://master-api.flutter.dev/flutter/dart-ui/ErrorCallback.html since we have to return something but `_defaultOnError` can be null.
@ueman opinions?

Comment (denrase) [reply flutter/lib/src/integrations/on_error_integration.dart]: If a user provides their own handler which we access with `_defaultOnError`, shouldn't we interpret it as 'handled'? Or is this not a case that will happen?

Comment (ueman) [reply flutter/lib/src/integrations/on_error_integration.dart]: Nope, I've got nothing to add to what you said.

File: flutter/lib/src/integrations/flutter_error_integration.dart (hunk 2 above)
Comment (marandaneto) [reply flutter/lib/src/integrations/flutter_error_integration.dart]: If they provided `FlutterError.onError`, they most likely have overridden ours, so I'd still keep the same behavior.

File: flutter/lib/src/integrations/on_error_integration.dart (hunk 3 above)
Comment (marandaneto) [reply flutter/lib/src/integrations/on_error_integration.dart]: Same as https://github.com/getsentry/sentry-dart/pull/1186#discussion_r1047052173
I'd still treat it as `handled: false`, this is a Sentry flag and not related to Flutters forwarding error handling or not.
We just forward the call to `_defaultOnError` as a good citizen.
If the callback has to return the fallback of `true` or `false`, I am unsure, I'd dig into the default `ErrorCallback` in the Flutter framework and use what they use as a default value.

File: flutter/test/integrations/flutter_error_integration_test.dart [hunk 4]
```diff
@@ -23,7 +23,7 @@ void main() {
     }) {
       // replace default error otherwise it fails on testing
       FlutterError.onError =
-          handler ?? (FlutterErrorDetails errorDetails) async {};
+          handler ??= (FlutterErrorDetails errorDetails) async {};
```
Comment (marandaneto) [inline flutter/test/integrations/flutter_error_integration_test.dart]: This would only replace if theres nothing in there, which is different than the comment above

File: flutter/lib/src/integrations/on_error_integration.dart (hunk 3 above)
Comment (denrase) [reply flutter/lib/src/integrations/on_error_integration.dart]: I was not able to find and ocurence in `flutter` or `flutter/engine` where this callback is assigned a default implementation. Maybe i'm searching in the wrong place. The documentation mentioned that if false is returned, the error will be [printed to sdtout](https://master-api.flutter.dev/flutter/dart-ui/PlatformDispatcher/onError.html). So i'd suggest to return the value we had previously and treat the mechanisms handled separately (always set to false).

Comment (ueman) [reply flutter/lib/src/integrations/on_error_integration.dart]: The relevant code is here I believe https://github.com/flutter/engine/blob/74a9c7e0f9b16cf8e26b2d1ce7b7777ae2a07572/lib/ui/platform_dispatcher.dart#L1141-L1183

Comment (marandaneto) [reply flutter/lib/src/integrations/on_error_integration.dart]: Yep, return false as fallback

**Comment by marandaneto** [comment]
@denrase please check if the `_eventWithRemovedBreadcrumbsIfHandled` is not causing duplicated breadcrumbs as well because it depends on the `handled` flag and now its flipped.

**Comment by denrase** [comment]
@marandaneto Also removing them from unhandled but not crashing event mechanisms that we create in the SDK.

**Comment by marandaneto** [comment]
> @marandaneto Also removing them from unhandled but not crashing event mechanisms that we create in the SDK.

Not sure I understand this comment, mind elaborating?

Another pending point if not done yet, we have to test the integration with the Native SDKs and the release health feature.

**Comment by denrase** [comment]
@marandaneto Sorry, hope this is a better explanation. 

> Not sure I understand this comment, mind elaborating?

Ah sure, i was referring to the documentation of why we introduced the breadcrumb removal.

> We then remove the breadcrumbs in all cases but if it is handled == false, this is a signal that the app would crash and android would lose the breadcrumbs by the time the app is restarted to read the envelope.

My assumption is that we still need to remove the breadcrumbs in those new instances where we have set the value to `handled == false`, as the app is not crashing on flutter.

I have added unit tests that cover this new behaviour.

File: dart/lib/src/sentry_client.dart [hunk 5]
```diff
@@ -399,13 +399,24 @@ class SentryClient {
...
 
-    if (hasNoMechanism || hasOnlyHandledMechanism) {
+    final hasOnlyHandledMechanism = mechanisms
+        .every((e) => (e.handled ?? true));
+
+    final unhandledButNotCrashingMechanisms = <String>[
+      'PlatformDispatcher.onError',
+      'isolateError',
+      'runZonedGuarded',
+    ];
```
Comment (marandaneto) [inline dart/lib/src/sentry_client.dart]: @denrase I don't think we need to special case the mechanism names, only the`handled` matters.

Comment (denrase) [reply dart/lib/src/sentry_client.dart]: k, i removed this again

File: dart/test/event_processor/exception/io_exception_event_processor_test.dart [hunk 6]
```diff
@@ -69,6 +69,7 @@ void main() {
       );
       expect(event.exceptions?.first.mechanism?.type, 'OSError');
       expect(event.exceptions?.first.mechanism?.meta['errno']['number'], 54);
+      expect(event.exceptions?.first.mechanism?.handled, isNull);
```
Comment (marandaneto) [inline dart/test/event_processor/exception/io_exception_event_processor_test.dart:72]: You can undo this change in favor of https://github.com/getsentry/sentry-dart/pull/1216

File: dart/test/event_processor/exception/io_exception_event_processor_test.dart [hunk 7]
```diff
@@ -92,6 +93,7 @@ void main() {
       );
       expect(event.exceptions?.first.mechanism?.type, 'OSError');
       expect(event.exceptions?.first.mechanism?.meta['errno']['number'], 42);
+      expect(event.exceptions?.first.mechanism?.handled, isNull);
```
Comment (marandaneto) [inline dart/test/event_processor/exception/io_exception_event_processor_test.dart:96]: Same as https://github.com/getsentry/sentry-dart/pull/1186/files#r1064593230

File: CHANGELOG.md [hunk 8]
```diff
@@ -8,6 +8,10 @@
 - Feat: Use `Hint` for screenshots ([#1165](https://github.com/getsentry/sentry-dart/pull/1165))
 - Feat: Support custom units for custom measurements ([#1181](https://github.com/getsentry/sentry-dart/pull/1181))
 
+### Enhancements
+
+- Enha: Exceptions not handled by user code should be `unhandled: true` #1186 ([#1186](https://github.com/getsentry/sentry-dart/pull/1186))
```
Comment (marandaneto) [inline CHANGELOG.md]:
```suggestion
- Exceptions not handled by user code should be `unhandled: true` #1186 ([#1186](https://github.com/getsentry/sentry-dart/pull/1186))
```

We don't need a prefix because we already have the headlines

File: dart/lib/src/sentry_client.dart [hunk 9]
```diff
@@ -399,19 +399,19 @@ class SentryClient {
...
-    }
+    // final mechanisms = (event.exceptions ?? [])
+    //     .map((e) => e.mechanism)
+    //     .whereType<Mechanism>();
+    // final hasNoMechanism = mechanisms.isEmpty;
+
+    // final hasOnlyHandledMechanism = mechanisms
+    //     .every((e) => (e.handled ?? true));
+
+    // if (hasNoMechanism || hasOnlyHandledMechanism) {
```
Comment (denrase) [inline dart/lib/src/sentry_client.dart:410]: @marandaneto Is this final or are you still in implementation process?

**Comment by denrase** [comment]
@marandaneto Raised an [issue](https://github.com/getsentry/sentry-cocoa/issues/2593) on the cocoa side.

File: dart/lib/src/sentry_client.dart (hunk 9 above)
Comment (marandaneto) [reply dart/lib/src/sentry_client.dart:410]: I was just testing and found out that this has to be fixed, hence the comments.

**Comment by marandaneto** [comment]
> @marandaneto Raised an [issue](https://github.com/getsentry/sentry-cocoa/issues/2593) on the cocoa side.

Thanks, let's pause this PR for a bit, I'm talking to the Apple folks about this issue, this is a blocker for us.

**Comment by marandaneto** [comment]
@denrase this is unblocked now, https://github.com/getsentry/sentry-cocoa/issues/2593

**Comment by marandaneto** [comment]
Closing in favor of https://github.com/getsentry/sentry-dart/pull/1535
